File: RecipeSharing/src/app/services/recipe/recipe.service.ts
import { HttpClient, HttpHeaders } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { BehaviorSubject, Observable, tap } from 'rxjs';
import { environment } from '../../../environments/environment';

@Injectable({
  providedIn: 'root'
})
export class RecipeService {

  private baseURL: string = environment.BASE_API_URL;

  constructor(private http: HttpClient) { }

  recipeSubject = new BehaviorSubject<any>({
    recipes: [],
    loading: false,
    newRecipe: null
  });

  private getHeaders(): HttpHeaders {
    const token = new HttpHeaders({
      'Content-Type': 'application/json',
      'Authorization': `Bearer ${localStorage.getItem('jwt')}`
    });
    return token;
  }

  getRecipes(): Observable<any> {
    const headers = this.getHeaders();
    return this.http.get(`${this.baseURL}/api/recipes`, { headers }).pipe(
      tap((recipes) => {
        const currentState = this.recipeSubject.value;
        this.recipeSubject.next({ ...currentState, recipes, loading: false });
      })
    );
}


  createRecipe(recipe: any): Observable<any> {
    const headers = this.getHeaders();
    return this.http.post(`${this.baseURL}/api/recipes`, recipe, { headers }).pipe(
      tap((newRecipe) => {
        const currentState = this.recipeSubject.value;
        this.recipeSubject.next({ ...currentState, recipes: [newRecipe, ...currentState.recipes] });
      })
    );
  }

  updateRecipe(recipe: any): Observable<any> {
    const headers = this.getHeaders();
    return this.http.put(`${this.baseURL}/api/recipes/update/${recipe.id}`, recipe, { headers }).pipe(
      tap((updatedRecipe: any) => {
        const currentState = this.recipeSubject.value;
        const updatedRecipes = currentState.recipes.map((item: any) => item.id === updatedRecipe.id ? updatedRecipe : item);
        this.recipeSubject.next({ ...currentState, recipes: updatedRecipes });
      })
    );
  }

  deleteRecipe(id: any): Observable<any> {
    const headers = this.getHeaders();
    return this.http.delete(`${this.baseURL}/api/recipes/${id}`, { headers, responseType: 'text' }).pipe(
      tap(() => {
        const currentState = this.recipeSubject.value;
        const updatedRecipes = currentState.recipes.filter((item: any) => item.id !== id);
        this.recipeSubject.next({ ...currentState, recipes: updatedRecipes });
      })
    );
  }
  

  likeRecipe(id: any): Observable<any> {
    const headers = this.getHeaders();
    return this.http.put(`${this.baseURL}/api/recipes/${id}/like`, {}, { headers }).pipe(
      tap((likedRecipe: any) => {
        const currentState = this.recipeSubject.value;
        const updatedRecipes = currentState.recipes.map((item: any) => item.id === likedRecipe.id ? likedRecipe : item);
        this.recipeSubject.next({ ...currentState, recipes: updatedRecipes });
      })
    );
  }

  getLikes(recipeId: any): Observable<any> {
    const headers = this.getHeaders();
    return this.http.get(`${this.baseURL}/api/recipes/${recipeId}/likes`, { headers }).pipe(
      tap((likes: any) => {
        const currentState = this.recipeSubject.value;
        const updatedRecipes = currentState.recipes.map((item: any) => {
          if (item.id === recipeId) {
            item.likes = likes;
          }
          return item;
        });
        this.recipeSubject.next({ ...currentState, recipes: updatedRecipes });
      })
    );
  }
  
}

File: recipedekhoapp/lib/services/recipe_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import 'package:shared_preferences/shared_preferences.dart';
import 'package:rxdart/rxdart.dart';

class RecipeService {
  final String baseUrl = 'https://knowrecipe-deployment-latest.onrender.com';

  final BehaviorSubject<Map<String, dynamic>> recipeSubject =
      BehaviorSubject.seeded({'recipes': [], 'loading': false});

  Future<String?> _getToken() async {
    final prefs = await SharedPreferences.getInstance();
    return prefs.getString('jwt');
  }

  Future<Map<String, String>> _getHeaders() async {
    final token = await _getToken();
    return {
      'Content-Type': 'application/json',
      'Authorization': 'Bearer $token',
    };
  }

  Future<List<dynamic>> getRecipes() async {
    recipeSubject.add({...recipeSubject.value, 'loading': true});
    try {
      final headers = await _getHeaders();
      final response = await http.get(Uri.parse('$baseUrl/api/recipes'), headers: headers);
      final List<dynamic> recipes = jsonDecode(response.body);
      recipeSubject.add({'recipes': recipes, 'loading': false});
      return recipes;
    } catch (e) {
      recipeSubject.add({...recipeSubject.value, 'loading': false});
      throw Exception('Error fetching recipes: $e');
    }
  }

  Future<dynamic> getRecipeById(String id) async {
    final headers = await _getHeaders();
    final response = await http.get(Uri.parse('$baseUrl/api/recipes/$id'), headers: headers);
    return jsonDecode(response.body);
  }

  Future<dynamic> createRecipe(Map<String, dynamic> recipe) async {
    final headers = await _getHeaders();
    final response = await http.post(Uri.parse('$baseUrl/api/recipes'), headers: headers, body: jsonEncode(recipe));
    return jsonDecode(response.body);
  }

  Future<dynamic> updateRecipe(Map<String, dynamic> recipe) async {
    final headers = await _getHeaders();
    final response = await http.put(Uri.parse('$baseUrl/api/recipes/update/${recipe["id"]}'), headers: headers, body: jsonEncode(recipe));
    return jsonDecode(response.body);
  }

  Future<dynamic> deleteRecipe(String id) async {
    final headers = await _getHeaders();
    final response = await http.delete(Uri.parse('$baseUrl/api/recipes/$id'), headers: headers);
    print(response.runtimeType);
    return response;
  }

  Future<dynamic> likeRecipe(String id) async {
    final headers = await _getHeaders();
    final response = await http.put(Uri.parse('$baseUrl/api/recipes/$id/like'), headers: headers);
    return jsonDecode(response.body);
  }

  Future<dynamic> getLikes(String recipeId) async {
    final headers = await _getHeaders();
    final response = await http.get(Uri.parse('$baseUrl/api/recipes/$recipeId/likes'), headers: headers);
    return jsonDecode(response.body);
  }
}
